### ISSUE TITLE
Race condition when async() within transaction

### ISSUE THREAD

**Issue by k4ml** [Koed00/django-q#154]
This has been explained in Celery docs - http://docs.celeryproject.org/en/latest/userguide/tasks.html#database-transactions

And we bumped into this issue after rolling out django-q into production recently. Let say we have a function like this:-

```python
@transaction.atomic
def send_msg(recipient, text, msgid=None, queue=False):
    msg = Message.objects.get_or_create(msgid=msgid)
    if queue:
        async('send_msg', recipient, text, msg.msgid, queue=False)
        msg.status = 'QUEUED'
        msg.save()
        return msg

    # pass msg to backend gateway
    send_to_gateway(recipient, msg, ....)

Once we had the qcluster running, we start seeing "Duplicate entry error" in the log. Look like when we call async(), the cluster picked it up before the transaction get committed, so get_or_create() there will create new object with the same msgid instead of getting an existing one. So we fix that using `transaction.on_commit()` in django 1.9.

```python
@transaction.atomic
def send_msg(recipient, text, msgid=None, queue=False):
    msg = Message.objects.get_or_create(msgid=msgid)
    if queue:
        def _add_to_queue():
            async('send_msg', recipient, text, msg.msgid, queue=False)

        msg.status = 'QUEUED'
        msg.save()
        transaction.on_commit(_add_to_queue)
        return msg

    # pass msg to backend gateway
    send_to_gateway(recipient, msg, ....)
```
Maybe we should have this somewhere in the docs, like in Celery docs.

**Comment by Koed00** [comment]
This is happening with the ORM broker?

**Comment by k4ml** [comment]
No, we use SQS but it doesn't matter what broker you use. It's depend whether your task depend on something in db. In our case above, the flow basically:-

Web request

1. User send message
1. We create a record in the db to represent the sent message with unique id
1. async() the function to send the message, with the message id as one of the arguments.
1. commit the transaction.

Queue worker

1. The send message function invoked from qcluster
1. Try to get the message by message id or create new one if doesn't exists.
1. Send the message and update status as SENT.
1. Commit transaction.

This issue happened between no. 3 in web request and no. 2 in the queue worker. Basically the rule of thumb if you're using transaction, async should be called from the outside of the transaction, not inside.

**Comment by Koed00** [comment]
Ok. Now I understand the problem. Thanks for explaining it like this.
Yes. We should add something about this to the docs, cause it can be a very complicated to debug.

**Comment by zmrenwu** [comment]
According to celery suggestion: http://docs.celeryproject.org/en/latest/userguide/tasks.html#database-transactions. doesn't work.
For now ,I just catch IntegrityError.

**Comment by k4ml** [comment]
@zmrenwu You mean using on_commit() function didn't work?
